### PR TITLE
Allow having multiple contacts with same role for one component

### PR DESCRIPTION
**Pull request by ycheng-aa**
Add a field count_limit to ContactRole to define the limit of
this role for each component.
The value default is 1, and define 0 as unlimited. When exceed
the limit, a ValidationError exception will be raised.

JIRA: PDC-1056

### CODE REVIEW

**Comment by lubomir** [comment]
Minor nitpick: the `contact-roles` API should should probably show the limit.

A bigger concern is that now the constraint is handled in the serializer. In the future, if we are creating contacts anywhere else in the code, the role count would not be checked. For this reason, I would prefer the check in `clean` method on the actual model. The check would be executed automatically by `pre_save` signal. It would even simplify the implementation as it would not need to handle create and update differently. In order not to have to duplicate code, a single mixin could implement the `clean` method and be used for both models.

**Comment by lubomir** [comment]
Looks pretty good to me now. Please add a period at the end of error message for consistency.

The code could simplified a bit in my opinion:
 * get all matching contacts
 * exclude current one if it has `pk`
 * perform count check

This way it would nicely fit into single method.

I also think the check for unlimited should be done before querying the database. It would save us one query in some cases.

File: pdc/apps/contact/models.py [hunk 1]
```diff
@@ -171,10 +186,10 @@ class ReleaseComponentContact(models.Model):
                                   on_delete=models.PROTECT)
 
     def __unicode__(self):
-        return u'%s: %s: %s' % (unicode(self.component), self.contact_role, unicode(self.contact))
+        return u'%s: %s: %s' % (unicode(self.component), self.role, unicode(self.contact))
```
Comment (erichuanggit) [inline pdc/apps/contact/models.py]: should add `unicode` for `role` as well?

File: pdc/apps/contact/models.py [hunk 2]
```diff
@@ -8,11 +8,14 @@
 from django.db import models
 from django.db.models.query import QuerySet
 from django.forms.models import model_to_dict
+from django.core.exceptions import ValidationError
 
 
 class ContactRole(models.Model):
 
     name = models.CharField(max_length=128, unique=True)
+    count_limit = models.IntegerField(default=1)
```
Comment (erichuanggit) [inline pdc/apps/contact/models.py]: how about use `count_contact_limit`? for my first glimpse, I understood it as the count limit of Role :-)

**Comment by erichuanggit** [comment]
Generally looks good to me.

File: pdc/apps/contact/models.py (hunk 2 above)
Comment (lubomir) [inline pdc/apps/contact/models.py]: Adding a comment seems better that changing the name. Maybe [help text](https://docs.djangoproject.com/en/dev/ref/models/fields/#help-text) or just mention in docstring would be enough?

File: pdc/apps/contact/models.py (hunk 1 above)
Comment (ycheng-aa) [inline pdc/apps/contact/models.py]: Fixed in0982fbd94

File: pdc/apps/contact/models.py (hunk 2 above)
Comment (ycheng-aa) [inline pdc/apps/contact/models.py]: Added help text in 0982fbd94.

**Comment by lubomir** [comment]
Looks good.

**Comment by ycheng-aa** [comment]
Merged.

**Pull request closed without merging**